### PR TITLE
Pipelines Pool Abstraction

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
This is an experiment in making pipelines pooling into a true abstraction, i.e. pipelines use MemoryPool<byte> instead of MemoryPool concrete class. The only usage of the concrete class is to create the default instances.

@davidfowl, @pakrym, @ahsonkhan, any thoughts?

### CODE REVIEW

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test Innerloop OSX10.12 Release Build and Test

File: src/System.IO.Pipelines/System/Buffers/MemoryPoolSlab.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@ namespace System.Buffers
     /// Slab tracking object used by the byte buffer memory pool. A slab is a large allocation which is divided into smaller blocks. The
     /// individual blocks are then treated as independant array segments.
     /// </summary>
-    public class MemoryPoolSlab : IDisposable
+    internal class MemoryPoolSlab : OwnedMemory<byte>
```
Comment (davidfowl) [inline src/System.IO.Pipelines/System/Buffers/MemoryPoolSlab.cs]: Why is this an OwnedMemory?

Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines/System/Buffers/MemoryPoolSlab.cs]: The allocation/deallocation callbacks need OwnedMemory. Since you call the callbacks when slabs are allocated/deallocated, I either needed to inherit slab from OwnedMemory or create a temp (throwaway) OwnedMemory when calling the callbacks. 

I also considered the callbacks to take something else than OwnedMemory (e.g. tuple of pointer/len), but I think the pool creates an illusion of OwnedMemory instances being the unit of allocation, and so OwnedMemory as a parameter of the callbacks seems to be consistent with the abstraction.

Comment (pakrym) [reply src/System.IO.Pipelines/System/Buffers/MemoryPoolSlab.cs]: Are callbacks only used in RIO sample, @davidfowl?

Comment (davidfowl) [reply src/System.IO.Pipelines/System/Buffers/MemoryPoolSlab.cs]: I'd rather remove the callbacks altogether and keep the slabs as is. Copy this pool as is into the RIO code and make this one internal.

Comment (davidfowl) [reply src/System.IO.Pipelines/System/Buffers/MemoryPoolSlab.cs]: @pakrym yes it is.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs [hunk 2]
```diff
@@ -10,12 +9,19 @@ namespace System.Buffers.Internal
...
+        const int DefaultSize = 4096;
+
         readonly static ArrayMemoryPool<T> s_shared = new ArrayMemoryPool<T>();
 
         public static ArrayMemoryPool<T> Shared => s_shared;
 
-        public override OwnedMemory<T> Rent(int minimumBufferSize)
+        public override int MaxBufferSize => 1024 * 1024 * 1024;
+
+        public override OwnedMemory<T> Rent(int minimumBufferSize = DefaultSize)
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: I don't think we should use default value here. Especially while having `AnySize` option.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs [hunk 3]
```diff
@@ -10,12 +9,19 @@ namespace System.Buffers.Internal
...
         {
+            if (minimumBufferSize == 0) minimumBufferSize = 32;
+            else if (minimumBufferSize == AnySize) minimumBufferSize = DefaultSize;
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: Why do we support both `AnySize` and `0`?

File: src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs [hunk 4]
```diff
@@ -5,9 +5,13 @@ namespace System.Buffers
...
     public abstract class MemoryPool<T> : IDisposable
     {
+        public const int AnySize = int.MinValue;
+
         public static MemoryPool<T> Default => Internal.ArrayMemoryPool<T>.Shared;
 
-        public abstract OwnedMemory<T> Rent(int minimumBufferSize);
+        public abstract OwnedMemory<T> Rent(int minimumBufferSize = AnySize);
+
+        public abstract int MaxBufferSize { get;  }
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: Why is this required?

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: Default and Any are different. Default is a const. Any is a sentinel. Any can return different sizes each time it's called. Default will return the same size.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: What would you prefer to happen when 0 is passed?

File: src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: So callers can query it to know that Rent won't throw (like your slab pool does)

Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: How would they react to the knowledge?

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs (hunk 3 above)
Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: We need to decide what const are we using for `AnySize` and throw in all other non-valid cases.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs (hunk 2 above)
Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: But it get's compiled into consumer code. So if implementation changes it would require a recompile to react.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: OK, I will change it to throw for zero

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: I don't get it: who needs to react to what? If the calling code does not pass anything to Rent, it apparently does not care what size it gets back. 

Having said that, I will change it to Any to be consistent across all pools.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs (hunk 4 above)
Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: Usually renting code lives deep inside the library and won't be able to do anything except to throw, but pool already does throwing.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs (hunk 2 above)
Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/ArrayMemoryPool.cs]: Imagine we have pipes `MemoryPool` with block size `4096` and `minimumBufferSize` being the same. Then we ship a new version with smaller block size. While user code looks like it doesn't care about the renting size it would start failing until rebuilt against a newer version of the library.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: I don't agree with "and won't be able to do anything" part. 

Also, see https://github.com/dotnet/corefxlab/blob/010238eb489ccfb04a47774cc606653afb82464a/src/System.IO.Pipelines.Networking.Sockets/SocketConnection.cs#L356

Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: Okay, I'm fine with keeping it, maybe consumer would be able to adjust to block size.

Comment (pakrym) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: What if there is no limit? `int.MaxValue`? Maybe it's better to make it nullable to make it more explicit?

Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: int.Max value is the limit, as OwnedMemory<T> cannot be larger.

**Comment by KrzysztofCwalina** [comment]
@pakrym, good to merge?

File: src/System.IO.Pipelines/System/Buffers/MemoryPoolSlab.cs [hunk 5]
```diff
@@ -20,7 +20,7 @@ public class MemoryPoolSlab : IDisposable
         private byte[] _data;
 
         private bool _isActive;
-        internal Action<MemoryPoolSlab> _deallocationCallback;
+        internal Action<OwnedMemory<byte>> _deallocationCallback;
```
Comment (davidfowl) [inline src/System.IO.Pipelines/System/Buffers/MemoryPoolSlab.cs]: These can be removed now.
